### PR TITLE
Closes #491: Make typescript renderers for PowerGrid package

### LINKED ISSUE #491: Make typescript renderers for PowerGrid package

### PR DESCRIPTION
**Pull request by tanas80**
TypeScript custom cell renderers using Rect and Canvas additional methods in js-api
+eslint rules

### CODE REVIEW

File: packages/PowerGrid/src/sparklines/piechart.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import * as DG from 'datagrok-api/dg';
+import * as ui from 'datagrok-api/ui';
+import {GridCell, GridColumn, Rect} from 'datagrok-api/src/grid';
+import {getSettingsBase, names, SummarySettingsBase} from './base';
+
+
+interface PiechartSettings extends SummarySettingsBase {
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/piechart.ts:7]: Should be "PieChartSettings"

File: packages/PowerGrid/src/cell-types/bar-cell-renderer.ts [hunk 2]
```diff
@@ -9,18 +9,23 @@ interface BarCellSettings {
...
   radius: DG.Property.js('radius', TYPE.STRING),
-}
+};
 
 export class BarCellRenderer extends DG.GridCellRenderer {
-  get name() { return 'bar'; }
-  get cellType() { return 'bar'; }
+  get name() {return 'bar';}
+
+  get cellType() {return 'bar';}
```
Comment (skalkin) [inline packages/PowerGrid/src/cell-types/bar-cell-renderer.ts:17]: Should be surrounded with spaces, like previously: `get cellType() { return 'bar'; }`

File: packages/PowerGrid/src/cell-types/bar-cell-renderer.ts [hunk 3]
```diff
@@ -9,18 +9,23 @@ interface BarCellSettings {
...
 
-  render(g: CanvasRenderingContext2D, x: number, y: number, w: number, h: number, gridCell: DG.GridCell, cellStyle: DG.GridCellStyle) {
+  render(
+    g: CanvasRenderingContext2D,
+    x: number, y: number, w: number, h: number,
+    gridCell: DG.GridCell, cellStyle: DG.GridCellStyle,
```
Comment (skalkin) [inline packages/PowerGrid/src/cell-types/bar-cell-renderer.ts:22]: Let's not end argument list with commas

File: packages/PowerGrid/src/sparklines/barchart.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+function getSettings(gc: DG.GridColumn): BarchartSettings {
+  return gc.settings ??= {
+    ...getSettingsBase(gc),
+    // ...{normalize: true},
+  };
+}
+
+export class BarchartCellRenderer extends DG.GridCellRenderer {
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/barchart.ts:18]: Should be "BarChartCellRenderer"

File: packages/PowerGrid/src/sparklines/barchart.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          .getLeftPart(cols.length, i)
+          .getBottomScaled(cols[i].scale(row))
+          .inflateRel(0.9, 1);
+
+        g.fillRect(bb.left, bb.top, bb.width, bb.height);
+      }
+    }
+
+    try {
+      // throw new Error("test");
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/barchart.ts:52]: What is the reason for this empty try-catch block?

File: packages/PowerGrid/src/sparklines/barchart.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  renderSettings(gc: GridColumn): Element {
+    gc.settings ??= getSettings(gc);
+    const settings = gc.settings;
+
+    return ui.inputs([
+      ui.columnsInput('Barchart columns', gc.grid.dataFrame, (columns) => {
+        settings.columnNames = names(columns);
+        gc.grid.invalidate();
+        console.log(JSON.stringify(gc.settings));
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/barchart.ts:72]: Let's get rid of debug outputs once the code is tested

File: packages/PowerGrid/src/sparklines/base.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+export function names(columns: Iterable<DG.Column>): string[] {
+  return Array.from(columns).map((c: any) => c.name);
+}
+
+export interface SummarySettingsBase {
+  columnNames: string[];
+}
+
+export function getSettingsBase(
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/base.ts:11]: The signature comfortably fits in one line, let's keep it simple

File: packages/PowerGrid/src/sparklines/base.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import * as DG from 'datagrok-api/dg';
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/base.ts:1]: If this file is supposed to contain common utilities, we usually call it "utils.ts" or "shared.ts"; I have nothing against "base.ts", in fact I like short names and it's a good one, but it's probably better to stick with the existing names

File: packages/PowerGrid/src/sparklines/piechart.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    for (let i = 0; i < cols.length; i++) {
+      if (cols[i].isNone(row))
+        continue;
+
+
+      const r = cols[i].scale(row) * box.width / 2;
+      g.beginPath();
+      g.moveTo(box.midX, box.midY);
+      g.arc(box.midX, box.midY, r,
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/piechart.ts:53]: We usually set default line width to 120 characters (80 would probably fine too, but you are probably using something even smaller)

File: packages/PowerGrid/src/sparklines/piechart.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    const settings = getSettings(gridCell.gridColumn);
+    const row: number = gridCell.cell.row.idx;
+    const cols = df.columns.byNames(settings.columnNames);
+    const box = new Rect(x, y, w, h).fitSquare().inflate(-2, -2);
+
+    for (let i = 0; i < cols.length; i++) {
+      if (cols[i].isNone(row))
+        continue;
+
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/piechart.ts:48]: We typically use two empty lines to separate classes, there is no reason to have them inside a function

File: packages/PowerGrid/src/sparklines/piechart.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      g.moveTo(box.midX, box.midY);
+      g.arc(box.midX, box.midY, r,
+        2 * Math.PI * i / cols.length, 2 * Math.PI * (i + 1) / cols.length);
+      g.closePath();
+
+      g.fillStyle = DG.Color.toRgb(DG.Color.getCategoricalColor(i));
+      g.fill();
+    }
+
+    /*
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/piechart.ts:61]: Let's get rid of commented out code once it's not useful anymore (and it looks like your renderers are working)

File: packages/PowerGrid/src/sparklines/radarchart.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    if (w < 20 || h < 10) return;
+
+    const settings = getSettings(gridCell.gridColumn);
+    const box = new Rect(x, y, w, h).fitSquare().inflate(-2, -2);
+    const row = gridCell.cell.row.idx;
+    const cols = df.columns.byNames(settings.columnNames);
+
+    g.strokeStyle = 'lightgray';
+
+    // axises' point calculator
```
Comment (skalkin) [inline packages/PowerGrid/src/sparklines/radarchart.ts:53]: plural for "axis" is "axes" :)

**Review by skalkin — CHANGES_REQUESTED** [review]
This looks good to me, let's just address a couple of the cosmetic issues I've mentioned, and let's merge it. Thanks! -Andrew

File: packages/PowerGrid/src/sparklines/piechart.ts (hunk 9 above)
Comment (tanas80) [reply packages/PowerGrid/src/sparklines/piechart.ts:53]: I am using IDE Code Style for TypeScript settings with Hard wrap at 120.
And ESLint rule 
    "max-len": [
      "error",
      120
    ],

**Pull request merged**